src/CoreWCF.Primitives/src/CoreWCF/Security/SecuritySessionServerSettings.cs

```diff
@@ -1893,42 +1867,92 @@ public bool TryReadSessionTokenIdentifier(XmlReader reader)
                 }
             }
 
-           //Renamed SecurityReplySessionChannel => SecurityReplySessionServiceChannelDispatcher (implementd IServiceChannelDispatcher)
-            public class SecurityReplySessionServiceChannelDispatcher : ServerSecuritySimplexSessionChannel, IServiceChannelDispatcher //, IReplySessionChannel 
+            //Renamed SecurityReplySessionChannel => SecurityReplySessionServiceChannelDispatcher (implementd IServiceChannelDispatcher)
+            public class SecurityReplySessionServiceChannelDispatcher : ServerSecuritySimplexSessionChannel, IServiceChannelDispatcher, IReplySessionChannel
```

Comment:
It seems like there might be some issues with the class hierarchy here. The original SecurityReplySessionChannel implementation in it's entirety it this:
```c#
        sealed class SecurityReplySessionChannel : SecurityReplyChannel, IReplySessionChannel
        {
            public SecurityReplySessionChannel(SecurityChannelListener<TChannel> channelManager, IReplySessionChannel innerChannel, SecurityProtocol securityProtocol, SecurityListenerSettingsLifetimeManager settingsLifetimeManager)
                : base(channelManager, innerChannel, securityProtocol, settingsLifetimeManager)
            {
            }

            public IInputSession Session
            {
                get { return ((IReplySessionChannel)this.InnerChannel).Session; }
            }
        }
```
It would seem to me that a base class called SecurityReplyServiceChannelDispatcher should exist which this class derives from which has most of the implementation. Looking at the WCF source code, SecurityReplyChannel has the code to deal with processing the message received and sending a reply fault if necessary. Too much deviation from the original WCF code makes me nervous about reasoning about equivalence of functionality.

Reply:
hey Matt you clicked the first link... (this class is the second link here => here is https://referencesource.microsoft.com/#System.ServiceModel/System/ServiceModel/Security/SecuritySessionServerSettings.cs,3494 ) and additional async methods are because we are changing to servicechannel dispatcher